File: extra_learning/design_patterns/07_chain_of_responsibility.py
"""Chain of responsibility / middleware."""
from __future__ import annotations
from collections.abc import Callable


def auth(next_: Callable[[str], None]) -> Callable[[str], None]:
    def inner(req: str) -> None:
        print("auth ok"); next_(req)
    return inner


def log_(next_: Callable[[str], None]) -> Callable[[str], None]:
    def inner(req: str) -> None:
        print("log:", req); next_(req)
    return inner


def handler(req: str) -> None:
    print("handler:", req)


if __name__ == "__main__":
    pipeline = log_(auth(handler))
    pipeline("hello")

File: go_for_data_engineering/05_design_patterns/07_chain_of_responsibility.go
// Chain of responsibility: middleware pattern, common in HTTP servers.
package main

import (
	"fmt"
)

type Middleware func(next func(string)) func(string)

func chain(mws ...Middleware) Middleware {
	return func(final func(string)) func(string) {
		for i := len(mws) - 1; i >= 0; i-- {
			final = mws[i](final)
		}
		return final
	}
}

func auth(next func(string)) func(string) {
	return func(s string) { fmt.Println("auth ok"); next(s) }
}
func log_(next func(string)) func(string) {
	return func(s string) { fmt.Println("log:", s); next(s) }
}

func main() {
	handler := chain(log_, auth)(func(s string) { fmt.Println("handler:", s) })
	handler("hello")
}
